File: examples/java/BFSExamples.java
import java.util.*;

public class BFSExamples {
    
    // 1. Binary Tree Level Order Traversal
    public List<List<Integer>> levelOrder(TreeNode root) {
        List<List<Integer>> result = new ArrayList<>();
        if (root == null) return result;
        
        Queue<TreeNode> queue = new LinkedList<>();
        queue.offer(root);
        
        while (!queue.isEmpty()) {
            int size = queue.size();
            List<Integer> level = new ArrayList<>();
            for (int i = 0; i < size; i++) {
                TreeNode node = queue.poll();
                level.add(node.val);
                if (node.left != null) queue.offer(node.left);
                if (node.right != null) queue.offer(node.right);
            }
            result.add(level);
        }
        return result;
    }
    
    // 2. Number of Islands
    public int numIslands(char[][] grid) {
        int count = 0;
        for (int i = 0; i < grid.length; i++) {
            for (int j = 0; j < grid[0].length; j++) {
                if (grid[i][j] == '1') {
                    bfs(grid, i, j);
                    count++;
                }
            }
        }
        return count;
    }
    
    private void bfs(char[][] grid, int i, int j) {
        Queue<int[]> queue = new LinkedList<>();
        queue.offer(new int[]{i, j});
        grid[i][j] = '0';
        
        int[][] dirs = {{0,1}, {1,0}, {0,-1}, {-1,0}};
        while (!queue.isEmpty()) {
            int[] curr = queue.poll();
            for (int[] dir : dirs) {
                int x = curr[0] + dir[0], y = curr[1] + dir[1];
                if (x >= 0 && x < grid.length && y >= 0 && y < grid[0].length && grid[x][y] == '1') {
                    grid[x][y] = '0';
                    queue.offer(new int[]{x, y});
                }
            }
        }
    }
    
    // 3. Rotting Oranges
    public int orangesRotting(int[][] grid) {
        Queue<int[]> queue = new LinkedList<>();
        int fresh = 0;
        
        for (int i = 0; i < grid.length; i++) {
            for (int j = 0; j < grid[0].length; j++) {
                if (grid[i][j] == 2) queue.offer(new int[]{i, j});
                else if (grid[i][j] == 1) fresh++;
            }
        }
        
        if (fresh == 0) return 0;
        
        int minutes = 0;
        int[][] dirs = {{0,1}, {1,0}, {0,-1}, {-1,0}};
        
        while (!queue.isEmpty()) {
            int size = queue.size();
            for (int i = 0; i < size; i++) {
                int[] curr = queue.poll();
                for (int[] dir : dirs) {
                    int x = curr[0] + dir[0], y = curr[1] + dir[1];
                    if (x >= 0 && x < grid.length && y >= 0 && y < grid[0].length && grid[x][y] == 1) {
                        grid[x][y] = 2;
                        queue.offer(new int[]{x, y});
                        fresh--;
                    }
                }
            }
            minutes++;
        }
        
        return fresh == 0 ? minutes - 1 : -1;
    }
    
    // 4. Word Ladder
    public int ladderLength(String beginWord, String endWord, List<String> wordList) {
        Set<String> wordSet = new HashSet<>(wordList);
        if (!wordSet.contains(endWord)) return 0;
        
        Queue<String> queue = new LinkedList<>();
        queue.offer(beginWord);
        int level = 1;
        
        while (!queue.isEmpty()) {
            int size = queue.size();
            for (int i = 0; i < size; i++) {
                String word = queue.poll();
                if (word.equals(endWord)) return level;
                
                char[] chars = word.toCharArray();
                for (int j = 0; j < chars.length; j++) {
                    char original = chars[j];
                    for (char c = 'a'; c <= 'z'; c++) {
                        chars[j] = c;
                        String newWord = new String(chars);
                        if (wordSet.contains(newWord)) {
                            wordSet.remove(newWord);
                            queue.offer(newWord);
                        }
                    }
                    chars[j] = original;
                }
            }
            level++;
        }
        return 0;
    }
    
    // 5. Minimum Depth of Binary Tree
    public int minDepth(TreeNode root) {
        if (root == null) return 0;
        
        Queue<TreeNode> queue = new LinkedList<>();
        queue.offer(root);
        int depth = 1;
        
        while (!queue.isEmpty()) {
            int size = queue.size();
            for (int i = 0; i < size; i++) {
                TreeNode node = queue.poll();
                if (node.left == null && node.right == null) return depth;
                if (node.left != null) queue.offer(node.left);
                if (node.right != null) queue.offer(node.right);
            }
            depth++;
        }
        return depth;
    }
    
    // 6. Open the Lock
    public int openLock(String[] deadends, String target) {
        Set<String> dead = new HashSet<>(Arrays.asList(deadends));
        Set<String> visited = new HashSet<>();
        
        if (dead.contains("0000")) return -1;
        
        Queue<String> queue = new LinkedList<>();
        queue.offer("0000");
        visited.add("0000");
        int steps = 0;
        
        while (!queue.isEmpty()) {
            int size = queue.size();
            for (int i = 0; i < size; i++) {
                String curr = queue.poll();
                if (curr.equals(target)) return steps;
                
                for (int j = 0; j < 4; j++) {
                    char c = curr.charAt(j);
                    String up = curr.substring(0, j) + (c == '9' ? '0' : (char)(c + 1)) + curr.substring(j + 1);
                    String down = curr.substring(0, j) + (c == '0' ? '9' : (char)(c - 1)) + curr.substring(j + 1);
                    
                    if (!dead.contains(up) && !visited.contains(up)) {
                        visited.add(up);
                        queue.offer(up);
                    }
                    if (!dead.contains(down) && !visited.contains(down)) {
                        visited.add(down);
                        queue.offer(down);
                    }
                }
            }
            steps++;
        }
        return -1;
    }
    
    // 7. Binary Tree Right Side View
    public List<Integer> rightSideView(TreeNode root) {
        List<Integer> result = new ArrayList<>();
        if (root == null) return result;
        
        Queue<TreeNode> queue = new LinkedList<>();
        queue.offer(root);
        
        while (!queue.isEmpty()) {
            int size = queue.size();
            for (int i = 0; i < size; i++) {
                TreeNode node = queue.poll();
                if (i == size - 1) result.add(node.val);
                if (node.left != null) queue.offer(node.left);
                if (node.right != null) queue.offer(node.right);
            }
        }
        return result;
    }
    
    // 8. Shortest Path in Binary Matrix
    public int shortestPathBinaryMatrix(int[][] grid) {
        if (grid[0][0] == 1) return -1;
        int n = grid.length;
        if (n == 1) return 1;
        
        Queue<int[]> queue = new LinkedList<>();
        queue.offer(new int[]{0, 0, 1});
        grid[0][0] = 1;
        
        int[][] dirs = {{-1,-1}, {-1,0}, {-1,1}, {0,-1}, {0,1}, {1,-1}, {1,0}, {1,1}};
        
        while (!queue.isEmpty()) {
            int[] curr = queue.poll();
            int x = curr[0], y = curr[1], dist = curr[2];
            
            for (int[] dir : dirs) {
                int nx = x + dir[0], ny = y + dir[1];
                if (nx == n - 1 && ny == n - 1) return dist + 1;
                if (nx >= 0 && nx < n && ny >= 0 && ny < n && grid[nx][ny] == 0) {
                    grid[nx][ny] = 1;
                    queue.offer(new int[]{nx, ny, dist + 1});
                }
            }
        }
        return -1;
    }
    
    // 9. Perfect Squares
    public int numSquares(int n) {
        Queue<Integer> queue = new LinkedList<>();
        Set<Integer> visited = new HashSet<>();
        queue.offer(n);
        visited.add(n);
        int level = 0;
        
        while (!queue.isEmpty()) {
            int size = queue.size();
            level++;
            for (int i = 0; i < size; i++) {
                int curr = queue.poll();
                for (int j = 1; j * j <= curr; j++) {
                    int next = curr - j * j;
                    if (next == 0) return level;
                    if (!visited.contains(next)) {
                        visited.add(next);
                        queue.offer(next);
                    }
                }
            }
        }
        return level;
    }
    
    // 10. Walls and Gates
    public void wallsAndGates(int[][] rooms) {
        Queue<int[]> queue = new LinkedList<>();
        for (int i = 0; i < rooms.length; i++) {
            for (int j = 0; j < rooms[0].length; j++) {
                if (rooms[i][j] == 0) queue.offer(new int[]{i, j});
            }
        }
        
        int[][] dirs = {{0,1}, {1,0}, {0,-1}, {-1,0}};
        while (!queue.isEmpty()) {
            int[] curr = queue.poll();
            for (int[] dir : dirs) {
                int x = curr[0] + dir[0], y = curr[1] + dir[1];
                if (x >= 0 && x < rooms.length && y >= 0 && y < rooms[0].length && 
                    rooms[x][y] == Integer.MAX_VALUE) {
                    rooms[x][y] = rooms[curr[0]][curr[1]] + 1;
                    queue.offer(new int[]{x, y});
                }
            }
        }
    }
    
    static class TreeNode {
        int val;
        TreeNode left, right;
        TreeNode() {}
        TreeNode(int val) { this.val = val; }
        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val; this.left = left; this.right = right;
        }
    }
    
    public static void main(String[] args) {
        BFSExamples examples = new BFSExamples();
        
        System.out.println("=== BFS Examples ===");
        
        // Test 1: Level Order Traversal
        TreeNode root = new TreeNode(3, new TreeNode(9), new TreeNode(20, new TreeNode(15), new TreeNode(7)));
        System.out.println("Level Order: " + examples.levelOrder(root));
        
        // Test 2: Number of Islands
        char[][] grid = {{'1','1','0','0','0'},{'1','1','0','0','0'},{'0','0','1','0','0'},{'0','0','0','1','1'}};
        System.out.println("Number of Islands: " + examples.numIslands(grid));
        
        // Test 3: Rotting Oranges
        int[][] oranges = {{2,1,1},{1,1,0},{0,1,1}};
        System.out.println("Rotting Oranges: " + examples.orangesRotting(oranges));
        
        // Test 4: Word Ladder
        System.out.println("Word Ladder: " + examples.ladderLength("hit", "cog", Arrays.asList("hot","dot","dog","lot","log","cog")));
        
        // Test 5: Minimum Depth
        System.out.println("Min Depth: " + examples.minDepth(root));
        
        // Test 6: Open the Lock
        System.out.println("Open Lock: " + examples.openLock(new String[]{"0201","0101","0102","1212","2002"}, "0202"));
        
        // Test 7: Right Side View
        System.out.println("Right Side View: " + examples.rightSideView(root));
        
        // Test 8: Shortest Path Binary Matrix
        int[][] matrix = {{0,0,0},{1,1,0},{1,1,0}};
        System.out.println("Shortest Path: " + examples.shortestPathBinaryMatrix(matrix));
        
        // Test 9: Perfect Squares
        System.out.println("Perfect Squares(12): " + examples.numSquares(12));
        
        // Test 10: Walls and Gates
        int[][] rooms = {{Integer.MAX_VALUE,-1,0,Integer.MAX_VALUE}};
        examples.wallsAndGates(rooms);
        System.out.println("Walls and Gates completed");
    }
}

File: examples/python/bfs_examples.py
from collections import deque

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

# 1. Level Order
def level_order(root):
    if not root:
        return []
    
    result = []
    queue = deque([root])
    
    while queue:
        level_size = len(queue)
        level = []
        for _ in range(level_size):
            node = queue.popleft()
            level.append(node.val)
            if node.left:
                queue.append(node.left)
            if node.right:
                queue.append(node.right)
        result.append(level)
    
    return result

# 2. Num Islands
def num_islands(grid):
    if not grid:
        return 0
    
    def bfs_islands(grid, i, j):
        queue = deque([(i, j)])
        grid[i][j] = '0'
        
        directions = [(0, 1), (1, 0), (0, -1), (-1, 0)]
        while queue:
            x, y = queue.popleft()
            for dx, dy in directions:
                nx, ny = x + dx, y + dy
                if 0 <= nx < len(grid) and 0 <= ny < len(grid[0]) and grid[nx][ny] == '1':
                    grid[nx][ny] = '0'
                    queue.append((nx, ny))
    
    count = 0
    for i in range(len(grid)):
        for j in range(len(grid[0])):
            if grid[i][j] == '1':
                bfs_islands(grid, i, j)
                count += 1
    return count

# 3. Oranges Rotting
def oranges_rotting(grid):
    queue = deque()
    fresh = 0
    
    for i in range(len(grid)):
        for j in range(len(grid[0])):
            if grid[i][j] == 2:
                queue.append((i, j))
            elif grid[i][j] == 1:
                fresh += 1
    
    if fresh == 0:
        return 0
    
    minutes = 0
    directions = [(0, 1), (1, 0), (0, -1), (-1, 0)]
    
    while queue:
        minutes += 1
        for _ in range(len(queue)):
            x, y = queue.popleft()
            for dx, dy in directions:
                nx, ny = x + dx, y + dy
                if 0 <= nx < len(grid) and 0 <= ny < len(grid[0]) and grid[nx][ny] == 1:
                    grid[nx][ny] = 2
                    queue.append((nx, ny))
                    fresh -= 1
    
    return minutes - 1 if fresh == 0 else -1

# 4. Ladder Length
def ladder_length(begin_word, end_word, word_list):
    word_set = set(word_list)
    if end_word not in word_set:
        return 0
    
    queue = deque([begin_word])
    level = 1
    
    while queue:
        for _ in range(len(queue)):
            word = queue.popleft()
            if word == end_word:
                return level
            
            for i in range(len(word)):
                for c in 'abcdefghijklmnopqrstuvwxyz':
                    new_word = word[:i] + c + word[i+1:]
                    if new_word in word_set:
                        word_set.remove(new_word)
                        queue.append(new_word)
        level += 1
    
    return 0

# 5. Min Depth
def min_depth(root):
    if not root:
        return 0
    
    queue = deque([root])
    depth = 1
    
    while queue:
        for _ in range(len(queue)):
            node = queue.popleft()
            if not node.left and not node.right:
                return depth
            if node.left:
                queue.append(node.left)
            if node.right:
                queue.append(node.right)
        depth += 1
    
    return depth

# 6. Open Lock
def open_lock(deadends, target):
    dead = set(deadends)
    if "0000" in dead:
        return -1
    
    queue = deque(["0000"])
    visited = {"0000"}
    steps = 0
    
    while queue:
        for _ in range(len(queue)):
            curr = queue.popleft()
            if curr == target:
                return steps
            
            for i in range(4):
                digit = int(curr[i])
                for d in [-1, 1]:
                    new_digit = (digit + d) % 10
                    new_combo = curr[:i] + str(new_digit) + curr[i+1:]
                    if new_combo not in dead and new_combo not in visited:
                        visited.add(new_combo)
                        queue.append(new_combo)
        steps += 1
    
    return -1

# 7. Right Side View
def right_side_view(root):
    if not root:
        return []
    
    result = []
    queue = deque([root])
    
    while queue:
        level_size = len(queue)
        for i in range(level_size):
            node = queue.popleft()
            if i == level_size - 1:
                result.append(node.val)
            if node.left:
                queue.append(node.left)
            if node.right:
                queue.append(node.right)
    
    return result

# 8. Shortest Path Binary Matrix
def shortest_path_binary_matrix(grid):
    if grid[0][0] == 1:
        return -1
    n = len(grid)
    if n == 1:
        return 1
    
    queue = deque([(0, 0, 1)])
    grid[0][0] = 1
    
    directions = [(-1,-1), (-1,0), (-1,1), (0,-1), (0,1), (1,-1), (1,0), (1,1)]
    
    while queue:
        x, y, dist = queue.popleft()
        for dx, dy in directions:
            nx, ny = x + dx, y + dy
            if nx == n - 1 and ny == n - 1:
                return dist + 1
            if 0 <= nx < n and 0 <= ny < n and grid[nx][ny] == 0:
                grid[nx][ny] = 1
                queue.append((nx, ny, dist + 1))
    
    return -1

# 9. Num Squares
def num_squares(n):
    queue = deque([n])
    visited = {n}
    level = 0
    
    while queue:
        level += 1
        for _ in range(len(queue)):
            curr = queue.popleft()
            i = 1
            while i * i <= curr:
                next_val = curr - i * i
                if next_val == 0:
                    return level
                if next_val not in visited:
                    visited.add(next_val)
                    queue.append(next_val)
                i += 1
    
    return level

# 10. Walls And Gates
def walls_and_gates(rooms):
    if not rooms:
        return
    
    queue = deque()
    for i in range(len(rooms)):
        for j in range(len(rooms[0])):
            if rooms[i][j] == 0:
                queue.append((i, j))
    
    directions = [(0, 1), (1, 0), (0, -1), (-1, 0)]
    while queue:
        x, y = queue.popleft()
        for dx, dy in directions:
            nx, ny = x + dx, y + dy
            if (0 <= nx < len(rooms) and 0 <= ny < len(rooms[0]) and 
                rooms[nx][ny] == float('inf')):
                rooms[nx][ny] = rooms[x][y] + 1
                queue.append((nx, ny))

def run_examples():
    print("=== BFS Examples ===")
    
    # Test 1: Level Order Traversal
    root = TreeNode(3, TreeNode(9), TreeNode(20, TreeNode(15), TreeNode(7)))
    print(f"Level Order: {level_order(root)}")
    
    # Test 2: Number of Islands
    grid = [['1','1','0','0','0'],['1','1','0','0','0'],['0','0','1','0','0'],['0','0','0','1','1']]
    print(f"Number of Islands: {num_islands(grid)}")
    
    # Test 3: Rotting Oranges
    oranges = [[2,1,1],[1,1,0],[0,1,1]]
    print(f"Rotting Oranges: {oranges_rotting(oranges)}")
    
    # Test 4: Word Ladder
    print(f"Word Ladder: {ladder_length('hit', 'cog', ['hot','dot','dog','lot','log','cog'])}")
    
    # Test 5: Minimum Depth
    print(f"Min Depth: {min_depth(root)}")
    
    # Test 6: Open the Lock
    print(f"Open Lock: {open_lock(['0201','0101','0102','1212','2002'], '0202')}")
    
    # Test 7: Right Side View
    print(f"Right Side View: {right_side_view(root)}")
    
    # Test 8: Shortest Path Binary Matrix
    matrix = [[0,0,0],[1,1,0],[1,1,0]]
    print(f"Shortest Path: {shortest_path_binary_matrix(matrix)}")
    
    # Test 9: Perfect Squares
    print(f"Perfect Squares(12): {num_squares(12)}")
    
    # Test 10: Walls and Gates
    rooms = [[float('inf'), -1, 0, float('inf')]]
    walls_and_gates(rooms)
    print("Walls and Gates completed")

if __name__ == "__main__":
    run_examples()